### ISSUE TITLE
Memory leak

### ISSUE THREAD

**Issue by LemADEC** [sheenrox82/TheMistsOfRioVII#10]
When connecting on a Vanilla server, RioV seems to keep registering invalid texture, causing memory leak over time:
```
[22:38:19] [Client thread/ERROR]: Couldn't render entity
net.minecraft.util.ReportedException: Registering texture
	at net.minecraft.client.renderer.texture.TextureManager.func_110579_a(SourceFile:91) ~[bqf.class:?]
	at sheenrox82.riovII.src.asm.HorseMethods.getHorseTexture(HorseMethods.java:55) ~[HorseMethods.class:?]
	at net.minecraft.client.renderer.entity.RenderHorse.func_110775_a(SourceFile:73) ~[bnv.class:?]
	at net.minecraft.client.renderer.entity.RenderHorse.func_110775_a(SourceFile:14) ~[bnv.class:?]
	at net.minecraft.client.renderer.entity.Render.func_110777_b(SourceFile:47) ~[bno.class:?]
	at net.minecraft.client.renderer.entity.RenderHorse.func_77036_a(SourceFile:50) ~[bnv.class:?]
	at net.minecraft.client.renderer.entity.RenderHorse.func_77036_a(SourceFile:14) ~[bnv.class:?]
	at net.minecraft.client.renderer.entity.RendererLivingEntity.func_76986_a(RendererLivingEntity.java:149) [boh.class:?]
	at net.minecraft.client.renderer.entity.RenderLiving.func_76986_a(SourceFile:23) [bok.class:?]
	at net.minecraft.client.renderer.entity.RenderLiving.func_76986_a(SourceFile:11) [bok.class:?]
	at net.minecraft.client.renderer.entity.RenderManager.func_147939_a(RenderManager.java:293) [bnn.class:?]
	at net.minecraft.client.renderer.entity.RenderManager.func_147936_a(RenderManager.java:271) [bnn.class:?]
	at net.minecraft.client.renderer.entity.RenderManager.func_147937_a(RenderManager.java:244) [bnn.class:?]
	at net.minecraft.client.renderer.RenderGlobal.func_147589_a(RenderGlobal.java:471) [bma.class:?]
	at net.minecraft.client.renderer.EntityRenderer.func_78471_a(EntityRenderer.java:1224) [blt.class:?]
	at net.minecraft.client.renderer.EntityRenderer.func_78480_b(EntityRenderer.java:1011) [blt.class:?]
	at net.minecraft.client.Minecraft.func_71411_J(Minecraft.java:990) [bao.class:?]
	at net.minecraft.client.Minecraft.func_99999_d(Minecraft.java:887) [bao.class:?]
	at net.minecraft.client.main.Main.main(SourceFile:148) [Main.class:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:1.7.0_60-ea]
	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source) ~[?:1.7.0_60-ea]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[?:1.7.0_60-ea]
	at java.lang.reflect.Method.invoke(Unknown Source) ~[?:1.7.0_60-ea]
	at net.minecraft.launchwrapper.Launch.launch(Launch.java:135) [launchwrapper-1.11.jar:?]
	at net.minecraft.launchwrapper.Launch.main(Launch.java:28) [launchwrapper-1.11.jar:?]
Caused by: java.lang.OutOfMemoryError
	at java.util.zip.Inflater.inflateBytes(Native Method) ~[?:1.7.0_60-ea]
	at java.util.zip.Inflater.inflate(Unknown Source) ~[?:1.7.0_60-ea]
	at java.util.zip.InflaterInputStream.read(Unknown Source) ~[?:1.7.0_60-ea]
	at java.util.jar.JarVerifier$VerifierStream.read(Unknown Source) ~[?:1.7.0_60-ea]
	at java.io.FilterInputStream.read(Unknown Source) ~[?:1.7.0_60-ea]
	at javax.imageio.stream.MemoryCache.loadFromStream(Unknown Source) ~[?:1.7.0_60-ea]
	at javax.imageio.stream.MemoryCacheImageInputStream.read(Unknown Source) ~[?:1.7.0_60-ea]
	at javax.imageio.stream.ImageInputStreamImpl.readFully(Unknown Source) ~[?:1.7.0_60-ea]
	at javax.imageio.stream.ImageInputStreamImpl.readFully(Unknown Source) ~[?:1.7.0_60-ea]
	at com.sun.imageio.plugins.png.PNGImageReaderSpi.canDecodeInput(Unknown Source) ~[?:1.7.0_60-ea]
	at javax.imageio.ImageIO$CanDecodeInputFilter.filter(Unknown Source) ~[?:1.7.0_60-ea]
	at javax.imageio.spi.FilterIterator.advance(Unknown Source) ~[?:1.7.0_60-ea]
	at javax.imageio.spi.FilterIterator.<init>(Unknown Source) ~[?:1.7.0_60-ea]
	at javax.imageio.spi.ServiceRegistry.getServiceProviders(Unknown Source) ~[?:1.7.0_60-ea]
	at javax.imageio.ImageIO.getImageReaders(Unknown Source) ~[?:1.7.0_60-ea]
	at javax.imageio.ImageIO.read(Unknown Source) ~[?:1.7.0_60-ea]
	at javax.imageio.ImageIO.read(Unknown Source) ~[?:1.7.0_60-ea]
	at net.minecraft.client.renderer.texture.LayeredTexture.func_110551_a(SourceFile:36) ~[bpt.class:?]
	at net.minecraft.client.renderer.texture.TextureManager.func_110579_a(SourceFile:72) ~[bqf.class:?]
	... 24 more
```

**Comment by Eternaldoom** [comment]
is this only on vanilla servers, or RioV servers too?

**Comment by LemADEC** [comment]
I've seen some memory leak on our test Forge server but didn't found who was the culprit yet on that environment.

**Comment by Eternaldoom** [comment]
Does the server have riov installed or not?

**Comment by sheenrox82** [comment]
LemADEC, answer the question correctly next time.

**Comment by LemADEC** [comment]
This is a client side log when connected on a vanilla server.
Vanilla implies there's no forge and no RioV.
The forge server does have RioV, but again, who cares, it's a client side leak.

You do realize bugs don't disappear by ignoring them, right?

**Comment by Eternaldoom** [comment]
Well, the way the horse armor is coded (the way with the least ASM) is incompatible with vanilla servers. A forge server with Riov, on the other hand, should be fine. Can you post the log from that?

**Comment by LemADEC** [comment]
Yes, working on it, already got a couple network leaks, still digging...
Any good tools to dump detailed memory usage so we can narrow it down faster?

**Comment by Eternaldoom** [comment]
Can you just post the log?

**Comment by sheenrox82** [comment]
Lol who said I'm ignoring the issue? We've been waiting for you to post data but you're just holding it off.

lol

**Comment by LemADEC** [comment]
Closing an issue before actually looking into it, that's ignoring it.
Not reopening an issue after admitting there's a problem you'll fix, isn't better in that matter.
Assuming I'm the same time zone, never sleep or work, that's disrespectful.
Ignoring my previous answer about Forge environment or a tool to do it faster, that's not good either.

Now back to the actual issue:
* when being idle for several hours in a map without mobs, single player world, creative with RioVII 1.1.1 in the pack => there's nothing in the logs (no entry at all), in F3, memory leak isn't perceptible. There's indeed a log jitter on memory use.
* in same setup, when spawning 59 horses, FPS drops heavily from 120 to less than 6 fps. Looking away from the horse, restore decent frame-rate. Also, memory consumption in F3 is moving way too fast.
* reloading the same setup without RioVII, FPS is the same whether looking at horse or not (120 fps). Also, memory consumption is fairly stable according to F3.

In both cases, there's absolutely no entries in the logs once the world is loaded, until I close it.
Later tonight, I'll try with 1.1.2 and no other mods, to see if there's a difference.

**Comment by sheenrox82** [comment]
How is me waiting for you to post shit "not looking into the issue"? Lol.

Sorry, I was sleeping because when you posted this above, it was 3 AM. I guess I'm disrespectful for getting sleep before school, to a random person on the internet.

**Comment by sheenrox82** [comment]
Hahaha, it's actually not RioV II causing the issues.
Just did a test on v1.1.2, and spawned a shit ton of horses, didn't lose any FPS, and there were no errors in the log.

http://i.imgur.com/Gv2BLEp.png 

Still maintaining a constant 50 - 60 FPS. Now, if you'll excuse me, I have better things to do, like school work. Or is me doing schoolwork disrespectful to you?

**Comment by LemADEC** [comment]
I've asked one of our staff to try it in single player with only RiovII installed.
Logs are here: http://pastebin.com/rj03V4uu
Context is Java 1.8.0_05 64-Bit, FML 7.10.85.1236 and RiovII 1.1.1
He created a fresh world, got 120 fps, spawned 30 horses, FPS dropped to 10 (see logs).

Next step is to try again with updated Forge and RiovII.

**Comment by sheenrox82** [comment]
In my post above, I am running Forge 10.13.2.1236, spawned well over thirty horses, and am running v1.1.2 with no FPS drop.

Also, the only evidence I see of an "FPS drop" is your friend saying, "30 horses spawned, 10 fps.". Which.... sorry to say this, isn't solid proof.

**Comment by LemADEC** [comment]
![2015-01-21_23 28 53](https://cloud.githubusercontent.com/assets/4776032/5846828/5f2b6746-a1c7-11e4-8754-698f20060517.png)
![2015-01-21_23 33 26](https://cloud.githubusercontent.com/assets/4776032/5846827/5f2b5152-a1c7-11e4-83fb-a274ab4a2971.png)
Here are 2 screenshots on same world with either just Forge 1286 or Forge 1286 + RiovII 1.1.2.
It's on my PC, using Java 1.7.0_60-ea 64bits

**Comment by sheenrox82** [comment]
Odd, because I don't get frame rate issues. Not to mention, when you said the memory usage would skyrocket when using RioV II when spawning horses, etc... the second picture is using more memory than with RioV II installed as well. :P

**Comment by LemADEC** [comment]
The memory is jumping around very fast.

**Comment by Eternaldoom** [comment]
I spawned maybe a hundred horses when testing the ASM and barely got any FPS lag.

**Comment by LemADEC** [comment]
Our staff guy confirmed issue with latest Forge and RiovII.
Any logs we could add to help narrow it down?

**Comment by sheenrox82** [comment]
Yes, THE log. So far, this entire time, you've supplied no actual proof of lagging besides those screenshots which I supplied proof of no lag before you even posted those. Eternaldoom claims to have no lag issues as well. I'm starting to think it's on your end.

**Comment by LemADEC** [comment]
I've posted screenshots with and without the lag since you don't want to believe us, as you requested; now you're still declining that we've the issue.
Instead of denying the issue, we can propose workaround, new logs or patches, or just hypothesis to work on.
For example:
- did you try in a debug or a clean environment? did you try with latest Forge and no other mods? with default configuration?
- did we try adjusting video settings?
- could we disable the horse modification with an option? what would be the impact?
- could we review the ASM code for potential issues or maybe new ideas? for example, are we sure on line 55 there's never a null value and all branches are valid? does it work properly when only a single texture is passed to the 'layered' function? https://github.com/sheenrox82/TheMistsOfRioVII/blob/37ee3013fe1e781da0dfebc31dfc537db5601401/java/sheenrox82/riovII/src/asm/HorseMethods.java#L55
- could we add some cache for those textures or do it differently? for example, only call it when there's an actual overlay to show?
- etc.

Basically, throw ideas, sort/prioritize them out and address them one by one.

**Comment by sheenrox82** [comment]
How am I denying the issue when I've been waiting for a proper log from you guys for a few days now? The only log we've gotten, is the partial one in the main issue post, which isn't enough detail, the log where one of your staff members says, "I am getting 10 fps." which isn't solid proof. Now you did post screenshots, and yes, you are lagging, but that doesn't the fact that Eternal and I are not getting lag, which I have proven. Now, if you continue to say I'm "denying the issue" when all I'm doing is waiting for god damn fucking proper log, then I WILL close this issue and lock it since you can't seem to ever post a log.

I tried in a clean environment, no extra mods, just Forge, with RioV in my /mods/ folder.

**Comment by BossLetsPlays** [comment]
@LemADEC I have solved this bug for you. There is an easy way to fix this on your end. There are two things you could try, 1: get a new computer that has more than an i3 dual core 0.5 ghz and also get a graphics card, or 2: host the server on a better computer that has internet speeds of more than 0.002 kbps

**Comment by Eternaldoom** [comment]
>does it work properly when only a single texture is passed to the 'layered' function?
No. The whole point is to make a _layered_ texture.

Also, the method is very similar to the way vanilla does it.

**Comment by sheenrox82** [comment]
Plus, I even SOOOOOOO MANY times in all of my mods, and basically everywhere, that I do not help without a full log of something. Which you have failed to deliver multiple times.

**Comment by LemADEC** [comment]
You've already got 3 logs and did nothing with it. We've already told you multiple times there's nothing in there and ask how to add more logs, none of those got answered so far.
@Eternaldoom when horse has no armor, the call is made with a single texture, or did I miss something?

**Comment by sheenrox82** [comment]
We've actually only got two, partial logs, not full logs. As you obviously did not read my statements, I said I need full logs of when the memory leaks happened, we keep asking for these, and yet you just never post them. Since you never post them, I cannot help. If there is "nothing in there" then where did you get the error that was posted in the original post of this issue? What do you mean "ask how to add more logs", when, you've already done it twice for partial logs. And as stated before, the second log of one of your friends saying he's "getting 10 fps" isn't enough info to go off of, since there is nothing else proving that something is wrong, just some random dude saying he's lagging.

**Comment by sheenrox82** [comment]
Since I've been trying to get you to post the full log of when the OutOfMemory error happened for the past FOUR DAYS, and you keep putting it off, thus making no progression in the resolution of this issue, I'm going to go ahead and close and lock this.
